docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,244 @@
+immediately.
+
+We could also consider storing the information that we were the losing LNS locally on the LNS. The
+advantages of that are:
+
+- subsequent Data (or Join) requests do not potentially switch the connection
+- same handling as in the Data flow (therefore easier to follow)
+  
+Not sure the following section is relevant as both data and join flow need DevEui.
+The problem is that the Data message flow relies on the DevAddr while here we only have the DevEui.
+Ways to solve this:
+
+- Split the creation of the LoRaDevice from fetching the twin, create a dummy LoRaDevice and
+  mark our LNS as the losing one.
+- Store outside of LoRaDevice a data-structure with DevEui: flag (winning or losing) and inject
+  also on the data message handler
+- Do nothing and accept that a one-off connection stealing on the first data message can happen.
+
+#### Delay on the LNS itself or on the Function ❔
+
+What are the scenarios that it's better that the function implements the delay?
+
+Disadvantages of using a Task.Delay on the Function:
+
+- Observability: potentially we are messing up the measurements of the Function duration.
+- Keeps the HTTP connection between the LNS-Function open for more time.
+  
+#### Device reset ❔
+
+What happens if we have a device reset between message A and B? We save the twin
+immediately (irrelevant but how does the Function get it for the deduplication decision?)
+
+Unclear what is the relationship between the twin and the cache on the function: in the cache we
+keep the framecounter down from the twin but not the fcnt up ❔
+
+- If both LNSs receive message B, LNS1 SHOULD reset the frame counter. LNS2 then
+call the Function first to reset the frame counter/clear the cache.
+  - The Function will update the device twin frame counter to 0 (it needs a DeviceClient that would
+    cause a temporary connection switch - to be investigated if this is a problem)
+  - Clear cache entry or update the entry with both frame counter down and up to 0.
+  - Returns the result to the LNS: whether it was the winning or losing one
+  - LNS reacts as described before
+
+Alternative would be to switch the order of the operations: first clear the cache and then (if we
+are the winning LNS) update the twin.
+
+#### Feature flag
+
+Unclear what is the exact use-case here ❔ We could still set the delay to 0.
+
+The stickiness feature can be disabled with a feature flag on the LNS configuration.
+
+- By default LNS implements sticky affinity. When the flag is set, LNS allow the connection
+  ping-pong to happen.
+- This flag is only checked if we are on the Drop deduplication strategy (as Mark and None do not
+  support stickiness anyway)
+
+#### Consequences / implementation
+
+- In memory storage is DevEui -> boolean.
+- Where to store locally on LNS❔ Can be on the
+
+  - LNS LoRaDeviceClient.ConnectionManager since all of the operations pass through it.
+  - LoRaDevice
```

Comment:
I think it will need to go to the LoRaDevice. But that's too far into the implementation detail for the ADR I think.

Reply:
Removed this section (but this comment might as well apply to the full ADR, though given the different code paths I think we ought to be specific and more technical in the ADR)